clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -85,12 +87,16 @@
     private int drainIndex;
     private final TransactionManager transactionManager;
 
+    // A per-partition queue of batches ordered by creation time for tracking the in-flight batches
+    private final ConcurrentMap<TopicPartition, List<ProducerBatch>> inFlightBatches;
```

Comment:
It seems to me that we need to synchronize on accesses to the contained `List<ProducerBatch>`. This list is updated in `maybeRemoveFromInflightBatches`, which called both from the sender background thread and from the network client threads. This unsafe access can corrupt the list in the present patch.

Reply:
I am not fully convinced that there is concurrent access to `List<ProducerBatch>`.  In KafkaProducer, there is only one background I/O thread (`Sender`) that turns the records into requests and transmit them to the cluster.  `Sender.run(long now)` --> `NetworkClient.poll` --> `NetworkClient.handle...` --> `Sender.handleProduceResponse` -->  `Sender.completeBatch` --> `Sender.reenqueueBatch` --> `RecordAccumulator.reenque` --> `RecordAccumulator.maybeRemoveFromInflightBatches`  is the call stack.   `NetworkClient` uses `Selector` for non-blocking i/o multiplexing, but it is not a thread. 

We need to synchronize access on `Deque<ProducerBatch> deque = entry.getValue();` as  both the producer thread and the sender thread can concurrently access `deque` at the same time.  Meanwhile, `inFlightBatches` is only being modified by `RecordAccumulator` methods that are called by `Sender` thread.

Reply:
I couldn't find the concurrent access either, so it may be ok. It actually makes me wonder why we are using a `ConcurrentMap`? To be honest, it would be more intuitive for the inflight batches to be tracked inside `Sender`. I'd suggest doing that if it's straightforward, but I'm fine leaving it here if it takes a lot of work.

Reply:
Yea. I think I was mistaken. If the callbacks from the network client are happening in the context of the sender thread, then there is no concurrent access on `inflightBatches`.

Reply:
updated the change to move `inFlightBatches` to `Sender`,  and changed its type from `ConcurrentMap` to `Map`.